### ISSUE TITLE
tests dev

### ISSUE THREAD

**Issue by shakfu** [shedskin/shedskin#324]
I'm creating an umbrella issue to aggregate discussion of problems/solutions related the `test2` folder which is still a work-in-progress.

Current major issue:

- [ ] test2 cmake machinery does not support multiple modules (a -> b -> c -> ...) while the builtin shedskin makefile build does.

**Comment by shakfu** [comment]
Note that the above issue is fixed in [PR325](https://github.com/shedskin/shedskin/pull/325)

**Comment by shakfu** [comment]
One of the major pending items besides including more tests is implementing windows support with the `cmake` build system for the `testst2` directory.

The objective is to target MSVC first. This means incorporating the followings compilation flags (as per shedskin FLAGS) into the cmake build system:

```make
CC=cl

CCFLAGS=$(CPPFLAGS) /MP /O2 /EHsc /TP /MD /Zi /nologo /W0  /Oi /Oy- /GL /Gm- /GS  \
/fp:precise /Zc:wchar_t /Zc:forScope /D "and"="&&" /D "or"="||" /D "not"="!" /D "WIN32" \
/D "_CONSOLE" /D "ALL_INTERIOR_POINTERS" /D "GC_NOT_DLL" /D "_UNICODE" \
/D "PCRE_STATIC" /I$(EXTRA_INCLUDE)

LFLAGS= $(LDFLAGS) /link /libpath:$(EXTRA_LIB) gc.lib pcre.lib user32.lib \
  /NODEFAULTLIB:"libcmt" /subsystem:console  /OPT:REF /OPT:ICF
```

The other requirement is building `gc.lib` and `pcre.lib` (user32.lib is provided by Windows I think).

**Comment by srepmub** [comment]
unfortunately I don't remember much of the old windows support, or how libraries were built.. :S

**Comment by shakfu** [comment]
It seems to me that to at least starting testing windows support for `shedskin`, one requires two things:

1. cross-platform build system
2. access to the source or binaries of shedskin's dependencies (libgc and libpcre)

The first requirement is addressed nicely by `cmake` which is is available as one of two methods used in the `tests2` directory. 

The second is more tricky, and initially I thought of including the source code of each of the dependencies by some automatic method and then compiling them into the build somehow. But this is not so straightforward if the source is unix-centric and not necessarily intended to be included as a git submodule.

Happily this second problem is already solved by the [conan](https://conan.io) cross-platform `c/c++` package manager which itself is a `pip`-installable python package.

These are the steps I've used to to obtain the dependencies in windows:

```bash
pip install conan
```

Then search for the dependencies using the [conan center](https://conan.io/center/) search engine. It should be easy to find `pcre` and `bdwgc/8.2.2` and `pcre/8.45`.

After reading the [use it](https://conan.io/center/bdwgc?tab=useit) section in the `bdwgc/8.2.2` entry, I came up with the following `conanfile.txt` in the `tests2` directory:

```ini
[requires]
bdwgc/8.2.2
pcre/8.45

[generators]
CMakeDeps
CMakeToolchain
```

Then to install and build the dependencies on windows system for the `tests2` directory using the cmake testing method:

```bash
cd build
conan install --build=missing ..
```

The above command will make the dependencies available to be used in `cmake`. 

`conan` works well with `cmake`, and there are various options to include the dependencies in the usual `CMakeLists.txt` file. 

These will be covered later once further testing is done on a windows machine.
